package junit.tests;

import junit.extensions.TestSetup;
import junit.framework.AssertionFailedError;
import junit.framework.Test;
import junit.framework.TestCase;
import junit.framework.TestResult;
import junit.framework.TestSuite;
import junit.tests.extensions.ExtensionTest.TornDown;
import junit.tests.framework.NoArgTestCaseTest;

/**
 * A helper test case for testing whether the testing method
 * is run.
 */
public class WasRun extends TestCase {
	public boolean fWasRun= false;
		protected void runTest() {
			fWasRun= true;
		}
		public void testRunningErrorInTestSetup() {
			TestCase test= new TestCase("failure") {
				public void runTest() {
					fail();
				}
			};
		
			TestSetup wrapper= new TestSetup(test);
		
			TestResult result= new TestResult();
			wrapper.run(result);
			assertTrue(!result.wasSuccessful());
		}
		public void testRunningErrorsInTestSetup() {
			TestCase failure= new TestCase("failure") {
				public void runTest() {
					fail();
				}
			};
		
			TestCase error= new TestCase("error") {
				public void runTest() {
					throw new Error();
				}
			};
		
			TestSuite suite= new TestSuite();
			suite.addTest(failure);
			suite.addTest(error);
			
			TestSetup wrapper= new TestSetup(suite);
		
			TestResult result= new TestResult();
			wrapper.run(result);
		
			assertEquals(1, result.failureCount());
			assertEquals(1, result.errorCount());
		}
		public void testSetupErrorInTestSetup() {
			WasRun test= new WasRun();
		
			TestSetup wrapper= new TestSetup(test) {
				public void setUp() {
					fail();
				}
			};
		
			TestResult result= new TestResult();
			wrapper.run(result);
		
			assertTrue(!test.fWasRun);
			assertTrue(!result.wasSuccessful());
		}
		public void testSetupErrorDontTearDown() {
			WasRun test= new WasRun();
		
			TornDown wrapper= new TornDown(test) {
				public void setUp() {
					fail();
				}
			};
		
			TestResult result= new TestResult();
			wrapper.run(result);
		
			assertTrue(!wrapper.fTornDown);
		}
		public void testNamelessTestCase() {
			TestCase t= new TestCase() {};
			try {
				t.run();
				fail();
			} catch (AssertionFailedError e) {
			}
		}
		public void testRunAndTearDownFails() {
			TornDown fails= new TornDown() {
				protected void tearDown() {
					super.tearDown();
					throw new Error();
				}
				protected void runTest() {
					throw new Error();
				}
			};
			verifyError(fails);
			assertTrue(fails.fTornDown);
		}
		public void testNoArgTestCasePasses() {
			Test t= new TestSuite(NoArgTestCaseTest.class);
			TestResult result= new TestResult();
			t.run(result);
			assertTrue(result.runCount() == 1);
			assertTrue(result.failureCount() == 0);
			assertTrue(result.errorCount() == 0);
		}
}